#10874: Enable test cases for concurrent instances in CCL all gather

Aswinmcw opened the pull request:
Enable test cases for concurrent instances in CCL all gather
Tracking Issue : #10874 

T3K Nightly Test CI - https://github.com/tenstorrent/tt-metal/actions/runs/10483348824
T3K Frequent Test - https://github.com/tenstorrent/tt-metal/actions/runs/10492084492
All Post Commit Test - https://github.com/tenstorrent/tt-metal/actions/runs/10493261845

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
#10874: Enable test cases for concurrent instances

SeanNijjar commented:
Posting info from our offline discussion to the PR for reference:
- new line all-gather tests ported from ring all-gather should be moved to a new nightly list for line allgather
- Adopt APIs presented in the PR for new tests https://github.com/tenstorrent/tt-metal/pull/10872

Aswinmcw committed:
#10874: Move test to separate file

Aswinmcw committed:
#10874: Add sweep test in new infra

SeanNijjar commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,178 @@
+TIMEOUT = 30
+
+
+def is_unsupported_case(input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout):
```
Please import this from the base test_all_gather file if this is a duplicate of that one

SeanNijjar commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,178 @@
+
+    input_tensor = torch.rand(input_shape).bfloat16()
+
+    input_tensors = torch.chunk(input_tensor, num_devices, dim)
```
This is not using the proper `ShardTensor2dMesh` API as outlined in this PR: https://github.com/tenstorrent/tt-metal/pull/10872. Please update this and all future tests to use this API as it's the official recommendation way to describe these tensors now.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,335 @@
+import itertools
+
+
+def is_unsupported_case(input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout):
```
import from base test_all_gather

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,335 @@
+
+    input_tensor = torch.rand(input_shape).bfloat16()
+
+    input_tensors = torch.chunk(input_tensor, num_devices, dim)
```
Same API usage comment

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,335 @@
+    ],
+)
+@pytest.mark.parametrize("enable_async", [True, False])
+def test_line_all_gather_on_t3000_post_commit(
```
What's the difference between this and the same function in test_all_gather?

Aswinmcw replied:
This def is using `t3k_device_mesh` fixture instead of old `all_devices`

SeanNijjar replied:
Sounds good. Please rename this to "_nightly_"

Aswinmcw committed:
#10874: Use t3k_device_mesh fixture and remove reused code

Aswinmcw committed:
#10874: Use t3k_device_mesh fixture for concurrent instances

Aswinmcw committed:
#10874: Use t3k_device_mesh fixture in sweep test

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+    ],
+)
+@pytest.mark.parametrize("enable_async", [True, False])
+def test_line_all_gather_on_t3000_post_commit_instances(
```
rename to nightly

SeanNijjar commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,110 @@
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    # for device in devices:
```
delete

SeanNijjar commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,110 @@
+            eq, output = comp_pcc(tt_output_tensor, input_tensor)
+        if not eq:
+            logger.error(f"output mismatch for tensor {i}")
+        # assert eq, f"{i} FAILED: {output}"
```
delete

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+    if is_known_failure:
+        pytest.skip(f"Skipping unsupported case {message}.")
+
+    # for device in devices:
```
delete

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+@pytest.mark.parametrize(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR), # https://github.com/tenstorrent/tt-metal/issues/9686
```
Please remove the issue comments as they do not apply here.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
+        # ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.L1),
```
Please re-enable this for nightly

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+    "input_dtype",
+    [
+        ttl.tensor.DataType.BFLOAT16,
+        # ttl.tensor.DataType.BFLOAT8_B, # https://github.com/tenstorrent/tt-metal/issues/9686
```
please re-enable

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+    if is_known_failure:
+        pytest.skip(f"Skipping unsupported case {message}.")
+
+    # devices = get_devices_for_t3000(all_devices, num_devices)
```
delete

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+@pytest.mark.parametrize(
+    "num_devices, num_instances, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR), # https://github.com/tenstorrent/tt-metal/issues/9686
```
please remove the issue part of the comment.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,273 @@
+    [
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
```
remote issue comment

SeanNijjar reviewed (changes requested):
Just minor fixes. Also please the post commit, t3000 unit tests, and t3000 frequent pipelines

Aswinmcw committed:
#10874: Fix test ncalls

jdesousa-TT commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,106 @@
+}
+
+
+def skip(
```
If this is to be used with the new sweep framework, please change this skip function to use the `invalidate_vector` syntax described here: https://docs.google.com/presentation/d/1zo1pS40IZAcVFivAUSJIYljjVerGWJccpVJsP5KL00E/edit#slide=id.g2ec533b0c78_0_82

jdesousa-TT replied:
Please update this function to be called `invalidate_vector`

jdesousa-TT commented on tests/sweep_framework/sweeps/line_all_gather.py:
```diff
@@ -0,0 +1,106 @@
+# This is the run instructions for the test, defined by the developer.
+# The run function must take the above-defined parameters as inputs.
+# The runner will call this run function with each test vector, and the returned results from this function will be stored.
+def run(
```
Required inputs to this run function must match the defined parameters above, make sure all inputs are defined in the parameters list so the test can run with the sweeps infra.

jdesousa-TT replied:
For the device setup, please see this new PR and accompanying README change. This will allow you to define the device mesh you need for your test: https://github.com/tenstorrent/tt-metal/pull/11486

You should copy the device fixture code you require to a generator function called `device_mesh_fixture` in this file. The sweep infra will pick this up, use it, and pass it to the test in the `device` parameter which you should include instead of t3k_device_mesh here.

Aswinmcw replied:
Thanks for the PR, I will look into it and try adding device generator

SeanNijjar commented:
Just checked with @jdesousa-TT, as of about `e12cs07`, test case generation and reporting should work from within the Austin cloud. Can you please take these changes and generate the new cases and get the initial success rate? Then we can start to merge.

Aswinmcw committed:
#10874: Add symlink for CI

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
I think this should be `test_all_gather_nightly`- the special line entrypoint will eventually be deleted and all all-gather invocations will go through a generic `all-gather` op (line and ring).

Additionally, the nightly tests from `test_all_gather.py` need to be moved into this file as well.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+from ttnn import ShardTensorToMesh
+
+
+def run_line_all_gather(
```
This run function should probably just go in the base test_all_gather file and replace the corresponding function there. The callers of this function can be updated to use the t3k_device_mesh. Then this should be imported to here.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+    for device in t3k_device_mesh.get_devices():
+        t3k_device.append(device)
+
+    t3000_device_rows = [
```
This will need to be updated to use the tensor sharding APIs but we can do that in a separate PR

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+    input_tensor_mesh = ttnn.to_device(ttnn_tensor, t3k_device_mesh)
+
+    result_mesh_tensors = []
+    for i, devices in enumerate(t3000_device_rows):
```
This loop and the one following should be wrapped in  a loop over `num_iters`

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+        result_mesh_tensors.append(tt_out_tensor)
+
+    ## Wait for completion
+    for i, devices in enumerate(t3000_device_rows):
```
this is the second loop

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+        # (8, 1, [8, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR),
+        # # (8, 1, [8, 1, 256, 32], 0, ttl.tensor.Layout.TILE),
+        # (8, 1, [8, 8, 256, 384], 1, ttl.tensor.Layout.ROW_MAJOR),
+        # (4, 2, [8, 8, 256, 384], 1, ttl.tensor.Layout.TILE),
```
Is there a reason these other 4-chip cases aren't enabled? The 8chip cases being disabled makes sense (and they should just be removed from the list), but all 4-chip cases here should be enabled.

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:
```diff
@@ -0,0 +1,263 @@
+@pytest.mark.parametrize(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR),
```
why is it commented out?

Aswinmcw committed:
#0: Minor change

Aswinmcw committed:
#10874: Enable cases

Aswinmcw committed:
#10874: Use ttnn calls

Aswinmcw committed:
#10874: Use loops

Aswinmcw committed:
#10874: Use deprecated version

Aswinmcw commented on tests/ttnn/unit_tests/operations/test_all_gather.py:
```diff
@@ -476,6 +530,71 @@ def test_line_all_gather_on_t3000_post_commit(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        (4, 2, [1, 4, 32, 3584], 1, ttnn.TILE_LAYOUT),
+        # (8, 1, [1, 8, 32, 2048], 1, ttnn.TILE_LAYOUT),
```
Done

SeanNijjar reviewed (changes requested):
Just need these last couple changes. Please also confirm your changes were picked up by the nightly pipeline. I don't see the necessary change listed here.

Please add the file to the nightly pipeline. You can search for reduce_scatter_nightly invocation to see where to invoke the new test file changes.

Also confirm t3000 frequent and post commit are passing. Post the pipeline links in the PR description.

Aswinmcw commented:
@SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed

SeanNijjar commented:
> @SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed https://github.com/tenstorrent/tt-metal/actions/runs/10483348824

Please also run post commit (always) and t3000 frequent (since some of the test code was modified for that as well.

Aswinmcw commented:
> > @SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed https://github.com/tenstorrent/tt-metal/actions/runs/10483348824
> 
> Please also run post commit (always) and t3000 frequent (since some of the test code was modified for that as well.

Both all post commit test and t3k frequent test are passing, i will ensure remaining tests before merging

Aswinmcw committed:
#10874: Modify sweep test to use device fixture

jdesousa-TT reviewed (approved):
Looks good to me now. Thanks.
